docs/rfcs/0003-rfc-controller_refactor.md

```diff
@@ -0,0 +1,99 @@
+
+## Motivation
+The current controller code is complicated, lengthy, and bug-prone. Over time we have continued to directly add more logic and checks for various scenarios to the controller code, resulting in a Frankenstein monster that is hard to scale and keep under control.
+
+Because our current approach is not scalable, we’ve been facing numerous issues in the connectivity space, especially when it comes down to identifying exactly what is causing the issue. This means the behavior of our client is extremely inconsistent both on different platforms but also within the same environment. For example, if a user loses network connectivity while the VPN is activated, sometimes they enter no signal, and other times they encounter the “server location unavailable” modal. This inconsistency is a result of racing checks and error reporting. The goal of this refactor is to combine all of these scenarios under one component and to get rid of the `ConnectionHealth` entirely to resolve this issue.
+
+
+## Proposed Solution
+Move all connectivity code out of the controller and into its separate component which is referred to as the `ConnectionManager` in this document. The controller object will continue to remain but its purpose will be drastically simplified to a few basic functionalities including `activate()` and `deactivate()`. We will no longer have any states associated with the Controller and all states will live in the `ConnectionManager`. The `ConnectionManager` will primarily act as a liaison between the controller and the UI. 
+
+```mermaid
+graph TD
+    ConnectionManager --> Frontend
+    Frontend --> ConnectionManager
+    ConnectionManager -->|activate, deactivate| Controller
+```
+
+The existing Connection Health object in VPN performs checks that act in the same space as the `ConnectionManager`. For example, in connection health, we periodically send pings and collect some data upon receiving (or not receiving) a response which allows us to draw conclusions regarding network connectivity and server availability. Because there is so much overlap within both classes, the creation and implementation of the Connection Manager component allows us to entirely get rid of connection health and move any remaining logic to `ConnectionManager`. We should be able to entirely eliminate the `ConnectionHealth` class as well as the states `NoSignal` and `Stable` given that they will be handled within the new Connection Manager logic. For example, currently, we may enter the No Signal state due to several reasons such as losing network connectivity, expired subscription, or server location becoming unavailable, this means that No Signal is simply a side effect of something else going wrong, and because we now have explicit messaging and checks for those, No Signal does not serve us and can be removed.
```

Comment:
This paragraph discusses the functionality of the ConnectionManager, but the only introduction to it is "Move all connectivity code out of the controller and into its separate component which is referred to as the `ConnectionManager`". What about including a section where the concrete features of the ConnectionManager are discussed? Nothing in depth necessary, but just a bridge to situate the reader.

Reply:
With the latest edits I think this is much clearer. We can resolve this one.